File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public static int maxDepth;
    public static int target;
    public static int cnt;
    public static void log(Object o){
        System.out.println(o);
    }
    public void dfs(int[] numbers, int depth, int sum){
        if(depth == maxDepth){
            if(sum == target){
                cnt += 1;
            }
            return;
        }
        // 현재숫자를 다음 재귀에 반영
        dfs(numbers, depth+1, sum + numbers[depth]);
        dfs(numbers, depth+1, sum - numbers[depth]);
    }
    public int solution(int[] numbers, int target) {
        // 더하거나 빼거나 해서 마지막에 도달했을 때 합을 구해서 target과 비교
        maxDepth = numbers.length;
        int answer = 0;
        this.target = target;
        dfs(numbers, 0, 0);
        log(cnt);
        return cnt;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
'''
더하기 or 빼기 2가지 선택지
2^20 = 1000 * 1000 = 1,000,000 

더하거나 빼기 둘 중에 하나를 선택하여 깊이가 n에 도달 했을 때 target을 비교하고 카운팅

'''

cnt = 0

def dfs(max_depth, cur_depth, numbers, tmp, target):
    global cnt
    if max_depth == cur_depth:
        #print(tmp)
        if tmp == target:
            cnt += 1
        return
    dfs(max_depth, cur_depth + 1, numbers, tmp + numbers[cur_depth], target)
    dfs(max_depth, cur_depth + 1, numbers, tmp - numbers[cur_depth], target)

def solution(numbers, target):
    global cnt
    n = len(numbers)
    dfs(n, 0, numbers, 0, target)
    #print("총 도달 카운트", cnt)
    return cnt